packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    /*
+     * @internal
+     */
+    @observable
+    public readonly slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRowData = unknown> {
+    // This method will produce the expected TableCellState that the cellTemplate expects as
+    // its source
+    generateCellState: (cellData: TCellData) => TableCellState<TCellData>;
+
+    // The template to use to render the cell content for the column
+    cellTemplate: ViewTemplate<TableCellState<TCellData>>;
+
+    // The keys that should be present in TCellData.
+    // This array is parallel with the keys returned from `getDataKeys()`.
+    readonly cellDataKeyNames: readonly string[];
+
+    // The keys from the row data that correlate to the data that will be in TCellData.
+    // This array is parallel with the keys specified by `dataKeyNames`.
+    getRowDataKeys: () => string[];
+
+    // Function that allows the table column to validate the type that gets created
+    // for the cell data. This should validate that the types in TCellData are correct
+    // for each key defined by `dataKeyNames`.
+    // Return `true` if the data is valid. Return `false` if the data is not valid.
+    validateCellData(cellData: TableRowData): boolean;
+
+    // The content to use for this particular table column. The 'header' slot is the
+    // default slot for a TableColumn.
+    @observable
+    readonly slottedHeader: HTMLElement;
```

Comment:
I'm not sure how to reflect this in the spec, but I think we said that the header might ultimately be slotted into the table rather than the column. I'm not sure we need to figure out the detail of how this would work within this spec, but from my understanding, it might be that `slottedHeader` does not exist on a `TableColumn`.

Reply:
This has been removed.